icaruscode/Filters/FilterDataIntegrity_module.cc

```diff
@@ -0,0 +1,96 @@
+////////////////////////////////////////////////////////////////////////
+//
+// FilterDataIntegrity class
+//
+////////////////////////////////////////////////////////////////////////
+#include <fstream>
+
+/// Framework includes
+#include "art/Framework/Core/ModuleMacros.h"
+#include "art/Framework/Core/EDFilter.h"
+#include "art/Framework/Principal/Event.h"
+#include "canvas/Utilities/InputTag.h"
+#include "fhiclcpp/ParameterSet.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+#include "lardataobj/RecoBase/Hit.h"
+
+#include "sbndaq-artdaq-core/Overlays/ICARUS/PhysCrateFragment.hh"
+
+///filters for events, etc
+namespace filter
+{
+
+    class FilterDataIntegrity : public art::EDFilter
+    {
+
+    public:
+        explicit FilterDataIntegrity(fhicl::ParameterSet const &);
+
+        bool filter(art::Event &evt) override;
+
+    private:
+
+    // We will use this to keep track of the expected fragments in an event
+    // but note we will assume the first event is complete to set this list
+    std::set<int> fExpectedFragments;
+
+    }; //class FilterDataIntegrity
+}
+
+///////////////////////////////////////////////////////
+
+filter::FilterDataIntegrity::FilterDataIntegrity(fhicl::ParameterSet const &pset)
+    : EDFilter{pset}
+{
+    return;
+}
+
+bool filter::FilterDataIntegrity::filter(art::Event &event)
+{
+    bool filterPass = true;
+
+	// get all the artdaq fragment collections in the event.
+	std::vector<art::Handle<std::vector<artdaq::Fragment>>> fragmentHandles;
+	fragmentHandles = event.getMany<std::vector<artdaq::Fragment>>();
+
+    std::set<int> missingFragments(fExpectedFragments);
+    int           emptyFragments(0);
+
+    for (const auto& handle : fragmentHandles)
+    {
+        for (const auto& fragment : *handle)
+        {
+            int fragmentID = fragment.fragmentID();
+
+            fExpectedFragments.insert(fragmentID);
+            missingFragments.erase(fragmentID);
+
+            std::string instanceName = handle.provenance()->productInstanceName();
+            std::size_t found        = instanceName.find("Empty");
+
+            if (found != std::string::npos)
+            {
+                // fragment is empty
+                emptyFragments++;
+            }
+        }
+    }
+
+    mf::LogDebug("FilterDataIntegrity") << "Expected fragments: " << fExpectedFragments.size() << " expected, has " << missingFragments.size() << " missing. There are " << emptyFragments << " empty fragments";
+
+    if (missingFragments.size() > 0 || emptyFragments > 0)
```

Comment:
Recommended the use of `empty()` to check whether a collection is empty:
```suggestion
    if (!missingFragments.empty() || emptyFragments > 0)
```

Reply:
Ok, in my haste I had forgotten if "empty" was universal or a properly of vectors only...

Reply:
In C++17 it has become an universal at the point where there is a free function `std::empty()`, so that in principle you should be able to use for any container a syntax like `empty(missingFragment)` (note the intentional absence of `std::`).